// @iobroker/js-controller-adapter/build/cjs/lib/adapter/defaultObjs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaultObjs_exports = {};
__export(defaultObjs_exports, {
  createDefaults: () => createDefaults
});
module.exports = __toCommonJS(defaultObjs_exports);
function createDefaults() {
  return {
    "level.dimmer": {
      type: "number",
      read: true,
      write: true,
      min: 0,
      max: 100,
      unit: "%"
    },
    "indicator.working": {
      def: false,
      type: "boolean",
      read: true,
      write: false
    },
    "indicator.maintenance": {
      def: false,
      type: "boolean",
      read: true,
      write: false
    },
    "indicator.maintenance.lowbat": {
      def: false,
      type: "boolean",
      read: true,
      write: false,
      desc: "Low battery"
    },
    "indicator.maintenance.unreach": {
      def: false,
      type: "boolean",
      read: true,
      write: false,
      desc: "Device unreachable"
    },
    switch: {
      type: "boolean",
      read: true,
      write: true
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createDefaults
});
