// @iobroker/js-controller-cli/build/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var esm_exports = {};
__export(esm_exports, {
  BackupRestore: () => import_setupBackup.BackupRestore,
  PacketManager: () => import_setupPacketManager.PacketManager,
  Upgrade: () => import_setupUpgrade.Upgrade,
  Upload: () => import_setupUpload.Upload,
  Vendor: () => import_setupVendor.Vendor,
  dbConnectAsync: () => import_dbConnection.dbConnectAsync,
  execute: () => import_setup.execute
});
module.exports = __toCommonJS(esm_exports);
var import_setup = require("./lib/setup.js");
var import_dbConnection = require("./lib/setup/dbConnection.js");
var import_setupVendor = require("./lib/setup/setupVendor.js");
var import_setupUpload = require("./lib/setup/setupUpload.js");
var import_setupUpgrade = require("./lib/setup/setupUpgrade.js");
var import_setupBackup = require("./lib/setup/setupBackup.js");
var import_setupPacketManager = require("./lib/setup/setupPacketManager.js");
__reExport(esm_exports, require("./lib/_Types.js"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BackupRestore,
  PacketManager,
  Upgrade,
  Upload,
  Vendor,
  dbConnectAsync,
  execute,
  ...require("./lib/_Types.js")
});
